contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,626 @@
+and the scheduler's internal cache. The two are **not guaranteed to be in sync**
+and extreme care should be taken when writing a plugin that uses data from both
+of them.
+
+Providing plugins access to the API server is necessary to implement useful
+features, especially when those features consume object types that the scheduler
+does not normally consider. Providing a `SharedInformerFactory` allows plugins
+to share caches safely.
+
+### Plugin Registration
+
+Each plugin must define a constructor and add it to the hard-coded registry. For
+more information about constructor args, see [Optional Args](#optional-args).
+
+Example:
+
+```go
+type PluginFactory = func(json.RawMessage, PluginHandle) (Plugin, error)
+
+type Registry map[string]PluginFactory
+
+func NewRegistry() Registry {
+   return Registry{
+      fooplugin.Name: fooplugin.New,
+      barplugin.Name: barplugin.New,
+      // New plugins are registered here.
+   }
+}
+```
 
-# WORK ESTIMATES
+It is also possible to add plugins to a `Registry` object and inject that into a
+scheduler. See [Custom Scheduler Plugins](#custom-scheduler-plugins-out-of-tree)
 
-We expect to see an early version of the scheduling framework in two release
-cycles (end of 2018). If things go well, we will start offering it as an
-alternative to the scheduler v1 by the end of Q1 2019 and start the deprecation
-of scheduler v1. We will make it the default scheduler of Kubernetes in Q2 2019,
-but we will keep the option of using scheduler v1 for at least two more release
-cycles.
-</s>
+## Plugin Lifecycle
+
+### Initialization
+
+There are two steps to plugin initialization. First,
+[plugins are registered](#plugin-registration). Second, the scheduler uses its
+configuration to decide which plugins to instantiate. If a plugin registers for
+multiple extension points, *it is instantiated only once*.
+
+When a plugin is instantiated, it is passed [config args](#optional-args) and a
+[`PluginHandle`](#pluginhandle).
+
+### Concurrency
+
+There are two types of concurrency that plugin writers should consider. A plugin
+might be invoked several times concurrently when evaluating multiple nodes, and
+a plugin may be called concurrently from *different
+[scheduling cycles](#scheduling-cycle)*.
+
+In the main thread of the scheduler, only one scheduling cycle is processed at a
+time. Any extension point up to and including [reserve](#reserve) will be
+finished before the next scheduling cycle begins*. After the reserve phase, the
+[permit](#permit) and [bind](#bind) phases are executed asynchronously. This
+means that a plugin could be called concurrently from two different scheduling
+cycles, provided that at least one of the calls is to an extension point after
+reserve. Stateful plugins should take care to handle these situations.
+
+Finally, [un-reserve](#un-reserve) plugins may be called from either the Permit
+thread or the Bind thread, depending on how the pod was rejected.
```

Comment:
Because Permit/Bind thread is async, un-reserve plugins will be called asynchronously. And it is possible that scheduler may schedule same pod (if pod is re-added into schedule queue, e.g. because pod updates) while pod is Permit/Bind thread. So pod may try to use reserved resources but these resources may be un-reserved in another thread later and can be used by another pod in next scheduling cycle.

e.g.

- (cycle 1) pod A reserves some resources
- (cycle 1) pod A enters into async Permit/Bind phase
- (cycle 2) scheduler tries to schedule pod A in main thread again
- (cycle 2) pod A reserves same resources because these resources are still reserved by it 
- (cycle 2) pod A enters into async Perm/Bind phase
- (cycle 1) pod A is denied in Permit or rejected in Bind phase, try to un-reserve resources
- (cycle 3) scheduler tries to scheduler pod B in main thread
- (cycle 3) pod B reserves same resources because they are un-reserved now

Now in cycle 2/3, pods A and B think they reserve same resources.

The root cause is it is possible to schedule same pod while it is in async Permit/Bind phase. Do you think is it necessary to avoid this in concurrent control?

Reply:
I think it would be a bug if we allowed pod A to be scheduled in cycle 2 while cycle 1 is not yet finished. More specifically, a pod should not re-enter the queue until un-reserve has happened.

Pod updates is an interesting case. Today I don't think a pod spec can change in a meaningful way, but I know that in-place resource updates are planned.

Reply:
Thanks for clarification!

FYI, in fixing a race in volume binder, I found a pod may be [re-added into scheduling queue](https://github.com/kubernetes/kubernetes/blob/d57d606275eb19074caad0f0dc1ab851755a34ca/pkg/scheduler/factory/factory.go#L569-L577) on pod updates. Note that [status update is not skipped](https://github.com/kubernetes/kubernetes/blob/d57d606275eb19074caad0f0dc1ab851755a34ca/pkg/scheduler/factory/factory.go#L401-L446), so it is possible that two pods will be in two different cycles (one is in async goroutine) for now.